packages/@aws-cdk/aws-ec2/lib/vpc-endpoint.ts

```diff
@@ -502,53 +502,69 @@ export class InterfaceVpcEndpoint extends VpcEndpoint implements IInterfaceVpcEn
+    }
+
+    // If we aren't going to lookup supported AZs we'll exit early, returning the subnetIds from the provided subnet selection
+    if (!lookupSupportedAzs) {
+      return subnets.map(s => s.subnetId);
+    }
 
-    // Some service names, such as AWS service name references, use Tokens to automatically
-    // fill in the region
-    // If it is an InterfaceVpcEndpointAwsService, then the reference will be resolvable since
-    // only references the region
+    // Some service names, such as AWS service name references, use Tokens to automatically fill in the region
+    // If it is an InterfaceVpcEndpointAwsService, then the reference will be resolvable since it only references the region
     const isAwsService = Token.isUnresolved(props.service.name) && props.service instanceof InterfaceVpcEndpointAwsService;
 
-    // Determine what name we pass to the context provider, either the verbatim name
-    // or a resolved version if it is an AWS service reference
-    let lookupServiceName = props.service.name;
-    if (isAwsService && !agnosticRegion) {
-      lookupServiceName = Stack.of(this).resolve(props.service.name);
-    } else {
-      // It's an agnostic service and we don't know how to resolve it.
-      // This is ok if the stack is region agnostic and we're not looking up
-      // AZs
-      lookupServiceName = props.service.name;
+    // Determine what service name gets pass to the context provider
+    // If it is an AWS service it will have a REGION token
+    const lookupServiceName = isAwsService ? Stack.of(this).resolve(props.service.name) : props.service.name;
+
+    // Check that the lookup will work
+    this.validateCanLookupSupportedAzs(subnets, lookupServiceName);
+
+    // Do the actual lookup for AZs
+    const availableAZs = this.availableAvailabilityZones(lookupServiceName);
+    const filteredSubnets = subnets.filter(s => availableAZs.includes(s.availabilityZone));
+
+    // Throw an error if the lookup filtered out all subnets
+    // VpcEndpoints must be created with at least one AZ
+    if (filteredSubnets.length == 0) {
+      throw new Error(`lookupSupportedAzs returned ${availableAZs} but subnets have AZs ${subnets.map(s => s.availabilityZone)}`);
     }
+    return filteredSubnets.map(s => s.subnetId);
+  }
+
+  /**
+   * Sanity checking when looking up AZs for an endpoint service, to make sure it won't fail
+   */
+  private validateCanLookupSupportedAzs(subnets: ISubnet[], serviceName: string) {
+
+    // Having any of these be true will cause the AZ lookup to fail at synthesis time
+    const agnosticAcct = Token.isUnresolved(this.stack.account);
+    const agnosticRegion = Token.isUnresolved(this.stack.region);
+    const agnosticService = Token.isUnresolved(serviceName);
+    // Having subnets with Token AZs can cause the endpoint to be created with no subnets, failing at deployment time
+    const agnosticSubnets = subnets.filter(s => Token.isUnresolved(s.availabilityZone));
```

Comment:
All the other ones are booleans. Do `subnets.some(s => Token.isUnresolves(...))` to turn this into a boolean as well.

----

Also add in: `Token.isUnresolved(subnets.map(s => s.availabilityZone))` as a check.

Motivation: check out this issue. The same thing that happened to subnetsIds over there could happen to availabilityZone Ids here (or convince yourself that it can't):

https://github.com/aws/aws-cdk/issues/4118#issuecomment-743311275

Reply:
@rix0rrr  the motiviation for `subnets.filter(s => Token.isUnresolved(s.availabilityZone))` was to be able to show the user *which* subnets are offending so they can debug them.